File: lib/aura/context/prompt/task_provider.rb
# frozen_string_literal: true

module Aura
  module Context
    class Prompt
      class TaskProvider
        def initialize(project_path)
          @project_path = project_path
        end

        def provide
          file = resolve_task_path
          return nil if file.nil? || file.empty?

          content = File.read(file, encoding: "utf-8").strip
          return nil if content.empty?

          "# LONG-RUN TASK\n#{content}"
        end

        private

        def resolve_task_path
          dir = File.expand_path(@project_path)
          while dir && dir != File.dirname(dir)
            file = File.join(dir, "task.md")
            return file if File.exist?(file)

            dir = File.dirname(dir)
          end
          nil
        end
      end
    end
  end
end

File: src/core/context/providers/taskProvider.ts
import fs from 'node:fs';
import path from 'node:path';

export class TaskProvider {
  private projectPath: string;

  constructor(projectPath: string) {
    this.projectPath = projectPath;
  }

  public provide(): string | null {
    const file = this.resolveTaskPath();
    if (!file) return null;

    try {
      const content = fs.readFileSync(file, 'utf-8').trim();
      if (!content) return null;
      return `# LONG-RUN TASK\n${content}`;
    } catch (_e) {
      return null;
    }
  }

  private resolveTaskPath(): string | null {
    let dir = path.resolve(this.projectPath);
    while (true) {
      const file = path.join(dir, 'task.md');
      if (fs.existsSync(file) && fs.statSync(file).isFile()) {
        return file;
      }
      const parent = path.dirname(dir);
      if (parent === dir) {
        break;
      }
      dir = parent;
    }
    return null;
  }
}
